File: solutions/cpp/Problem9.cpp
#pragma once
/*
9. Palindrome Number

Determine whether an integer is a palindrome. An integer is a palindrome when it reads the same backward as forward.
*/
#include <iostream>


class Solution {
public:
	bool isPalindrome(int x) {
		if (x < 0)
		{
			return false;
		}
		else
		{
			return (reverse(x) == x);
		}
	}
	int reverse(int x) {
		long long result = 0;
		while (x != 0)
		{
			result *= 10;
			result += x % 10;
			x /= 10;
		}
		if (result > INT_MAX || result < INT_MIN)
		{
			return 0;
		}
		return (int)result;
	}
	void test()	{
		std::cout << "isPalindrome 38466483: " << isPalindrome(38466483) << std::endl;
		std::cout << "isPalindrome -141: " << isPalindrome(-141) << std::endl;
		std::cout << "isPalindrome 2147483648: " << isPalindrome(2147483648) << std::endl;
	}
};

File: solutions/python3/problem9.py
# -*- coding: utf-8 -*-
"""
9. Palindrome Number

Determine whether an integer is a palindrome. An integer is a palindrome when it reads the same backward as forward.

Follow up:

Could you solve it without converting the integer to a string?
"""


class Solution:
    def isPalindrome(self, x: int) -> bool:
        if x < 0:
            return False
        elif x == 0:
            return True
        elif x % 10 == 0:
            return False
        else:
            tmp = 0
            while x > tmp:
                tmp = tmp * 10 + x % 10
                x = x // 10

            return x == tmp or x == tmp // 10
